File: cpp/2022-10-23/leetcode55.cpp
//
// Created by bossw on 2022/10/23.
//

#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    bool canJump(vector<int> &nums) {
        int target = nums.size() - 1;
        int i = target - 1;
        while (target > 0 && i >= 0) {
            if (i >= 0 && nums[i] >= (target - i)) {
                target = i;
                i = target - 1;
            } else
                i--;
        }
        if (target <= 0)
            return true;
        else
            return false;
    }
};

int main(void) {
    vector<int> nums{2, 0, 0};
    cout << (new Solution)->canJump(nums) << endl;
    return 0;
}

File: python/2022-10-23/leetcode55.py
from typing import List


class Solution:
    def canJump(self, nums: List[int]) -> bool:
        d = {}

        def dfs(i):
            if i == len(nums):
                return True
            if i > len(nums):
                d[i] = False
                return False
            for w in range(nums[i]):
                if d.get(i + w) is not None:
                    res = d.get(i - w)
                else:
                    res = dfs(i + w)
                if res:
                    break
            return res

        return dfs(0)


s = Solution
print(s.canJump(s, [3, 2, 1, 0, 4]))
